Snowflake.Data.Tests/IntegrationTests/SFConnectionIT.cs

```diff
@@ -1020,6 +1022,52 @@ public void TestSSOConnectionTimeoutAfter10s()
             Assert.LessOrEqual(stopwatch.ElapsedMilliseconds, (waitSeconds + 5) * 1000);
         }
 
+        [Test]
+        [Ignore("This test requires manual interaction and therefore cannot be run in CI")]
+        public void TestSSOConnectionWithTokenCaching()
+        {
+            using (IDbConnection conn = new SnowflakeDbConnection())
+            {
+                conn.ConnectionString
+                    = ConnectionStringWithoutAuth
+                        + ";authenticator=externalbrowser;user=qa@snowflakecomputing.com;allow_sso_token_caching=true;";
+
+                // Authenticate to retrieve and store the token if doesn't exist or invalid
+                conn.Open();
+                Assert.AreEqual(ConnectionState.Open, conn.State);
+
+                conn.Close();
+                Assert.AreEqual(ConnectionState.Closed, conn.State);
+
+                // Authenticate using the token
+                conn.Open();
+                Assert.AreEqual(ConnectionState.Open, conn.State);
+            }
+        }
+
+        [Test]
+        [Ignore("This test requires manual interaction and therefore cannot be run in CI")]
+        public void TestSSOConnectionWithInvalidCachedToken()
+        {
+            using (IDbConnection conn = new SnowflakeDbConnection())
+            {
+                conn.ConnectionString
+                    = ConnectionStringWithoutAuth
+                        + ";authenticator=externalbrowser;user=qa@snowflakecomputing.com;allow_sso_token_caching=true;";
+
+                var key = SFCredentialManagerFactory.BuildCredentialKey(testConfig.host, testConfig.user, TokenType.IdToken);
+                var credentialManager = SFCredentialManagerInMemoryImpl.Instance;
+                credentialManager.SaveCredentials(key, "wrongToken");
+
+                SFCredentialManagerFactory.SetCredentialManager(credentialManager);
+
+                conn.Open();
+                Assert.AreEqual(ConnectionState.Open, conn.State);
+
+                SFCredentialManagerFactory.UseDefaultCredentialManager();
+            }
+        }
+
```

Comment:
Is it possible to write a test which would not be ignored? Maybe with something mocked or something that simulates the browser action?

Reply:
I've added tests for external browser authentication using a mock browser